sql/core/src/test/scala/org/apache/spark/sql/CollationExpressionWalkerSuite.scala

```diff
@@ -0,0 +1,294 @@
+      case AbstractArrayType(elementType) => hasStringType(elementType)
+      case TypeCollection(typeCollection) =>
+        typeCollection.exists(hasStringType)
+      case StructType => true
+      case StructType(fields) => fields.exists(sf => hasStringType(sf.dataType))
+      case _ => false
+    }
+  }
+
+  def replaceExpressions(inputTypes: Seq[AbstractDataType], params: Seq[Class[_]]): Seq[Any] = {
+    (inputTypes, params) match {
+      case (Nil, mparams) => mparams
+      case (_, Nil) => Nil
+      case (minputTypes, mparams) if mparams.head.isAssignableFrom(classOf[Expression]) =>
+        minputTypes.head +: replaceExpressions(inputTypes.tail, mparams.tail)
+      case (minputTypes, mparams) =>
+        mparams.head +: replaceExpressions(minputTypes.tail, mparams.tail)
+    }
+  }
+
+  test("SPARK-48280: Expression Walker for Test") {
+    // This test does following:
+    // 1) Take all expressions
+    // 2) Find the ones that have at least one argument of StringType
+    // 3) Use reflection to create an instance of the expression using first constructor
+    //    (test other as well).
+    // 4) Check if the expression is of type ExpectsInputTypes (should make this a bit broader)
+    // 5) Run eval against literals with strings under:
+    //    a) UTF8_BINARY, "dummy string" as input.
+    //    b) UTF8_BINARY_LCASE, "DuMmY sTrInG" as input.
+    // 6) Check if both expressions throw an exception.
+    // 7) If no exception, check if the result is the same.
+    // 8) There is a list of allowed expressions that can differ (e.g. hex)
+    var expressionCounter = 0
+    var expectsExpressionCounter = 0;
+    val funInfos = spark.sessionState.functionRegistry.listFunction().map { funcId =>
+      spark.sessionState.catalog.lookupFunctionInfo(funcId)
+    }.filter(funInfo => {
+      // make sure that there is a constructor.
+      val cl = Utils.classForName(funInfo.getClassName)
+      !cl.getConstructors.isEmpty
+    }).filter(funInfo => {
+      expressionCounter = expressionCounter + 1
+      val cl = Utils.classForName(funInfo.getClassName)
+      // dummy instance
+      // Take first constructor.
+      val headConstructor = cl.getConstructors.head
+
+      val params = headConstructor.getParameters.map(p => p.getType)
+
+      val args = generateData(params.toSeq, Utf8Binary)
+      // Find all expressions that have string as input
+      try {
+        val expr = headConstructor.newInstance(args: _*)
+        expr match {
+          case types: ExpectsInputTypes =>
+            expectsExpressionCounter = expectsExpressionCounter + 1
+            val inputTypes = types.inputTypes
+            inputTypes.exists(hasStringType) || inputTypes.isEmpty
+        }
+      } catch {
+        case _: Throwable => false
+      }
+    }).toArray
+
+    val toSkip = List(
+      "parse_url", // Parse URL cannot be generalized with ExpectInputTypes
+      "hex" // Different inputs affect conversion
```

Comment:
how come that we didn't capture base64 and hashing expressions?

Reply:
I was surprised as well, but it seems that hashing function for UTF8_BINARY_LCASE was changed in CollationFactory to do .toLowerCase before hashing.

Reply:
```
return new Collation(
            "UTF8_BINARY_LCASE",
            PROVIDER_SPARK,
            null,
            UTF8String::compareLowerCase,
            "1.0",
            s -> (long) s.toLowerCase().hashCode(),
            /* supportsBinaryEquality = */ false,
            /* supportsBinaryOrdering = */ false,
            /* supportsLowercaseEquality = */ true);
```

Reply:
This was before ICU collation changes, please check https://github.com/apache/spark/pull/46180/files#diff-640c14aa5d7473df79b2435ce5a327dffcc16ca29354b153956b4f8d19fdb16cL237.

Reply:
Found it. Difference is that Hex can accept StringType, while base64 can only accept BinaryType, so base64 performs transformation to BinaryType before executing, which is using the hashing from CollationFactory, and hex is calling java's toBytes on the UTF8String. @dbatomic What behaviour is expected? I would say this is weird to have some conversions work one way and others the other way.